cmdstanpy/stanfit.py

```diff
@@ -1288,87 +1295,462 @@ def __repr__(self) -> str:
+                elif (
+                    inc_sample
+                    and var in self.mcmc_sample.metadata.stan_vars_cols
+                ):
+                    mcmc_vars.append(var)
+                else:
+                    raise ValueError('unknown variable: {}'.format(var))
+        else:
+            gq_cols = list(self.column_names)
+
+        if inc_sample and mcmc_vars:
+            if gq_cols:
+                return pd.concat(
+                    [
+                        self.mcmc_sample.draws_pd(
+                            vars=mcmc_vars, inc_warmup=inc_warmup
+                        ).reset_index(drop=True),
+                        pd.DataFrame(
+                            data=flatten_chains(
+                                self.draws(inc_warmup=inc_warmup)
+                            ),
+                            columns=self.column_names,
+                        )[gq_cols],
+                    ],
+                    axis='columns',
+                )
+            else:
+                return self.mcmc_sample.draws_pd(
+                    vars=mcmc_vars, inc_warmup=inc_warmup
+                )
+        elif inc_sample and vars is None:
+            cols_1 = self.mcmc_sample.column_names
+            cols_2 = self.column_names
+            dups = [
+                item
+                for item, count in Counter(cols_1 + cols_2).items()
+                if count > 1
+            ]
+            return pd.concat(
+                [
+                    self.mcmc_sample.draws_pd(inc_warmup=inc_warmup)
+                    .drop(columns=dups)
+                    .reset_index(drop=True),
+                    pd.DataFrame(
+                        data=flatten_chains(self.draws(inc_warmup=inc_warmup)),
+                        columns=self.column_names,
+                    ),
+                ],
+                axis='columns',
+                ignore_index=True,
+            )
+        elif gq_cols:
+            return pd.DataFrame(
+                data=flatten_chains(self.draws(inc_warmup=inc_warmup)),
+                columns=self.column_names,
+            )[gq_cols]
+
+        return pd.DataFrame(
+            data=flatten_chains(self.draws(inc_warmup=inc_warmup)),
+            columns=self.column_names,
         )
 
+    def draws_xr(
+        self,
+        vars: Union[str, List[str], None] = None,
+        inc_warmup: bool = False,
+        inc_sample: bool = False,
+    ) -> "xr.Dataset":
+        """
+        Returns the generated quantities draws as a xarray Dataset.
+        :param vars: optional list of variable names.
+        :param inc_warmup: When ``True`` and the warmup draws are present in
+            the MCMC sample, then the warmup draws are included.
+            Default value is ``False``.
+        """
+        if not XARRAY_INSTALLED:
+            raise RuntimeError(
+                "xarray is not installed, cannot produce draws array"
+            )
```

Comment:
I think I've found a way to test this, but I will open a separate PR after this is merged to address it. Testing something with an optional dependency can be very tricky